test/testrunner.py

```diff
@@ -38,21 +43,137 @@ def run_sagemaker_local_tests(images):
+    # Note: 3 is the max number of instances required for any tests. Here we schedule tests conservatively.
+    identifier = requester.send_request(image, "PR", 3)
+    image_tag = image.split(":")[-1]
+    report_path = os.path.join(os.getcwd(), "test", f"{image_tag}.xml")
+    while True:
+        query_status_response = requester.query_status(identifier)
+        test_status = query_status_response["status"]
+        if test_status == "completed":
+            LOGGER.info(f"Test for image {image} completed.")
+            logs_response = requester.receive_logs(identifier)
+            LOGGER.info(f"Receive logs success for ticket {identifier.ticket_name}, report path: {report_path}")
+            print_log_stream(logs_response)
+            metrics_utils.send_test_result_metrics(0)
+            with open(report_path, "w") as xml_report:
+                xml_report.write(logs_response["XML_REPORT"])
+            break
+
+        elif test_status == "runtimeError":
+            LOGGER.warning(f"Tests for image {image} ran into runtime error.")
+            logs_response = requester.receive_logs(identifier)
+            with open(report_path, "w") as xml_report:
+                xml_report.write(logs_response["XML_REPORT"])
+            print_log_stream(logs_response)
+            metrics_utils.send_test_result_metrics(1)
+            break
+
+        elif test_status == "failed":
+            LOGGER.warning(f"Scheduling failed. Reason: {query_status_response['reason']}")
+            metrics_utils.send_test_result_metrics(1)
+            break
+
+
 def run_sagemaker_remote_tests(images):
     """
     Function to set up multiprocessing for SageMaker tests
     :param images: <list> List of all images to be used in SageMaker tests
     """
-    if not images:
+    use_scheduler = os.getenv("USE_SCHEDULER", "False")
+    executor_mode = os.getenv("EXECUTOR_MODE", "False")
+
+    if executor_mode.lower() == "true":
+        LOGGER.info("entered executor mode.")
+        import log_return
+
+        num_of_instances = os.getenv("NUM_INSTANCES")
+        image = os.getenv("DLC_IMAGE")
+        job_type = "training" if "training" in image else "inference"
+        instance_type = sm_utils.assign_sagemaker_remote_job_instance_type(image)
+        test_succeeded = run_sagemaker_test_in_executor(image, num_of_instances, instance_type)
+
+        tag = image.split("/")[-1].split(":")[-1]
+        test_report = os.path.join(os.getcwd(), "test", f"{tag}.xml")
+
+        # update in-progress pool, send the xml reports
+        if test_succeeded:
+            log_return.update_pool("completed", instance_type, num_of_instances, job_type, test_report)
+        else:
+            log_return.update_pool("runtimeError", instance_type, num_of_instances, job_type, test_report)
         return
-    pool_number = len(images)
-    with Pool(pool_number) as p:
-        p.map(sm_utils.execute_sagemaker_remote_tests, images)
+
+    if use_scheduler.lower() == "true":
```

Comment:
Should we explicitly prevent executor_mode and use_scheduler from being True at the same time?

If the `return` in line 160 is removed, then it would be possible to send the scheduling into an infinite loop, right?